In [1]:
import pandas as pd

In [3]:
df = pd.read_csv('sales_data_final.csv')

In [4]:
df

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ...         category sub_category  \
0       Africa  ...  Office Supplies      Storage   
1      Oceania  ...  Office Supplies     Supplies   
2         EMEA  ...  Office Supplies      Storage   
3        North  ...  Office Supplies        Paper   
4      Oceania  ...        Furniture  Furnishings   
...        ...  ...              ...          ...   
51285     West  ...  Office Supplies      Binders   
51286   Africa  ...  Office Supplies      Binders   
51287  Central  ...  Office Supplies       Labels   
51288    North  ...  Office Supplies       Labels   
51289     West  ...  Office Supplies    Fasteners   

                                          product_name sales quantity  \
0                                  Tenex Lockers, Blue   408        2   
1                             Acme Trimmer, High Speed   120        3   
2                              Tenex Box, Single Width    66        4   
3                          Enermax Note Cards, Premium    45        3   
4                           Eldon Light Bulb, Duo Pack   114        5   
...                                                ...   ...      ...   
51285  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl    14        2   
51286          Wilson Jones Hole Reinforcements, Clear     4        1   
51287           Hon Color Coded Labels, 5000 Label Set    26        3   
51288           Hon Legal Exhibit Labels, Alphabetical     7        1   
51289                              Bagged Rubber Bands     3        3   

       discount    profit  shipping_cost  order_priority  year  
0           0.0  106.1400          35.46          Medium  2011  
1           0.1   36.0360           9.72          Medium  2011  
2           0.0   29.6400           8.17            High  2011  
3           0.5  -26.0550           4.82            High  2011  
4           0.1   37.7700           4.70          Medium  2011  
...         ...       ...            ...             ...   ...  
51285       0.2    4.5188           0.89          Medium  2014  
51286       0.0    0.4200           0.49          Medium  2014  
51287       0.0   12.3600           0.35          Medium  2014  
51288       0.0    0.5600           0.20          Medium  2014  
51289       0.2   -0.6048           0.17   

In [5]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year'],
      dtype='object')

In [7]:
df['order_date']

0          1/1/2011
1          1/1/2011
2          1/1/2011
3          1/1/2011
4          1/1/2011
            ...    
51285    12/31/2014
51286    12/31/2014
51287    12/31/2014
51288    12/31/2014
51289    12/31/2014
Name: order_date, Length: 51290, dtype: object

In [8]:
type(df['order_date'])

pandas.core.series.Series

In [9]:
type(df)

pandas.core.frame.DataFrame

In [10]:
df.dtypes

order_id           object
order_date         object
ship_date          object
ship_mode          object
customer_name      object
segment            object
state              object
country            object
market             object
region             object
product_id         object
category           object
sub_category       object
product_name       object
sales              object
quantity            int64
discount          float64
profit            float64
shipping_cost     float64
order_priority     object
year                int64
dtype: object

In [12]:
df[['profit','year']]

profit  year
0      106.1400  2011
1       36.0360  2011
2       29.6400  2011
3      -26.0550  2011
4       37.7700  2011
...         ...   ...
51285    4.5188  2014
51286    0.4200  2014
51287   12.3600  2014
51288    0.5600  2014
51289   -0.6048  2014

[51290 rows x 2 columns]

In [13]:
df['profit']  # data as series

0        106.1400
1         36.0360
2         29.6400
3        -26.0550
4         37.7700
           ...   
51285      4.5188
51286      0.4200
51287     12.3600
51288      0.5600
51289     -0.6048
Name: profit, Length: 51290, dtype: float64

In [14]:
df[['profit']]  # data as dataFrame

profit
0      106.1400
1       36.0360
2       29.6400
3      -26.0550
4       37.7700
...         ...
51285    4.5188
51286    0.4200
51287   12.3600
51288    0.5600
51289   -0.6048

[51290 rows x 1 columns]

In [15]:
type(df['profit'])

pandas.core.series.Series

In [16]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year'],
      dtype='object')

In [17]:
df[['order_id','product_id','quantity']]

order_id        product_id  quantity
0         AG-2011-2040  OFF-TEN-10000025         2
1        IN-2011-47883   OFF-SU-10000618         3
2         HU-2011-1220  OFF-TEN-10001585         4
3      IT-2011-3647632   OFF-PA-10001492         3
4        IN-2011-47883   FUR-FU-10003447         5
...                ...               ...       ...
51285   CA-2014-115427   OFF-BI-10002103         2
51286     MO-2014-2560  OFF-WIL-10001069         1
51287   MX-2014-110527   OFF-LA-10004182         3
51288   MX-2014-114783   OFF-LA-10000413         1
51289   CA-2014-156720   OFF-FA-10003472         3

[51290 rows x 3 columns]

In [18]:
df.describe() # return stats only of numerical columns

quantity      discount        profit  shipping_cost          year
count  51290.000000  51290.000000  51290.000000   51290.000000  51290.000000
mean       3.476545      0.142908     28.641740      26.375915   2012.777208
std        2.278766      0.212280    174.424113      57.296804      1.098931
min        1.000000      0.000000  -6599.978000       0.000000   2011.000000
25%        2.000000      0.000000      0.000000       2.610000   2012.000000
50%        3.000000      0.000000      9.240000       7.790000   2013.000000
75%        5.000000      0.200000     36.810000      24.450000   2014.000000
max       14.000000      0.850000   8399.976000     933.570000   2014.000000

In [19]:
# separate out columns based on data types

In [23]:
df.dtypes == 'object'

order_id           True
order_date         True
ship_date          True
ship_mode          True
customer_name      True
segment            True
state              True
country            True
market             True
region             True
product_id         True
category           True
sub_category       True
product_name       True
sales              True
quantity          False
discount          False
profit            False
shipping_cost     False
order_priority     True
year              False
dtype: bool

In [24]:
df.dtypes[df.dtypes == 'object']

order_id          object
order_date        object
ship_date         object
ship_mode         object
customer_name     object
segment           object
state             object
country           object
market            object
region            object
product_id        object
category          object
sub_category      object
product_name      object
sales             object
order_priority    object
dtype: object

In [25]:
df.dtypes[df.dtypes == 'object'].index

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'order_priority'],
      dtype='object')

In [26]:
df[df.dtypes[df.dtypes == 'object'].index]

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region        product_id         category sub_category  \
0       Africa  OFF-TEN-10000025  Office Supplies      Storage   
1      Oceania   OFF-SU-10000618  Office Supplies     Supplies   
2         EMEA  OFF-TEN-10001585  Office Supplies      Storage   
3        North   OFF-PA-10001492  Office Supplies        Paper   
4      Oceania   FUR-FU-10003447        Furniture  Furnishings   
...        ...               ...              ...          ...   
51285     West   OFF-BI-10002103  Office Supplies      Binders   
51286   Africa  OFF-WIL-10001069  Office Supplies      Binders   
51287  Central   OFF-LA-10004182  Office Supplies       Labels   
51288    North   OFF-LA-10000413  Office Supplies       Labels   
51289     West   OFF-FA-10003472  Office Supplies    Fasteners   

                                          product_name sales order_priority  
0                                  Tenex Lockers, Blue   408         Medium  
1                             Acme Trimmer, High Speed   120         Medium  
2                              Tenex Box, Single Width    66           High  
3                          Enermax Note Cards, Premium    45           High  
4                           Eldon Light Bulb, Duo Pack   114         Medium  
...                                                ...   ...            ...  
51285  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl    14         Medium  
51286          Wilson Jones Hole Reinforcements, Clear     4         Medium  
51287           Hon Color Coded Labels, 5000 Label Set    26         Medium  
51288           Hon Legal Exhibit Labels, Alphabetical     7         Medium  
51289                              Bagged Rubber Bands     3         Medium  

[51290 rows x 16 columns]

In [27]:
df[df.dtypes[df.dtypes == 'object'].index].describe()

order_id order_date   ship_date       ship_mode  \
count            51290      51290       51290           51290   
unique           25035       1430        1464               4   
top     CA-2014-100111  6/18/2014  11/22/2014  Standard Class   
freq                14        135         130           30775   

          customer_name   segment       state        country market   region  \
count             51290     51290       51290          51290  51290    51290   
unique              795         3        1094            147      7       13   
top     Muhammed Yedwab  Consumer  California  United States   APAC  Central   
freq                108     26518        2001           9994  11002    11117   

             product_id         category sub_category product_name  sales  \
count             51290            51290        51290        51290  51290   
unique            10292                3           17         3788   2246   
top     OFF-AR-10003651  Office Supplies      Binders      Staples     13   
freq                 35            31273         6152          227    589   

       order_priority  
count           51290  
unique              4  
top            Medium  
freq            29433

In [28]:
df.dtypes

order_id           object
order_date         object
ship_date          object
ship_mode          object
customer_name      object
segment            object
state              object
country            object
market             object
region             object
product_id         object
category           object
sub_category       object
product_name       object
sales              object
quantity            int64
discount          float64
profit            float64
shipping_cost     float64
order_priority     object
year                int64
dtype: object

In [29]:
df.dtypes[df.dtypes=='float'].index

Index(['discount', 'profit', 'shipping_cost'], dtype='object')

In [30]:
df[df.dtypes[df.dtypes=='float'].index]

discount    profit  shipping_cost
0           0.0  106.1400          35.46
1           0.1   36.0360           9.72
2           0.0   29.6400           8.17
3           0.5  -26.0550           4.82
4           0.1   37.7700           4.70
...         ...       ...            ...
51285       0.2    4.5188           0.89
51286       0.0    0.4200           0.49
51287       0.0   12.3600           0.35
51288       0.0    0.5600           0.20
51289       0.2   -0.6048           0.17

[51290 rows x 3 columns]

In [31]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year'],
      dtype='object')

In [32]:
df.order_id

0           AG-2011-2040
1          IN-2011-47883
2           HU-2011-1220
3        IT-2011-3647632
4          IN-2011-47883
              ...       
51285     CA-2014-115427
51286       MO-2014-2560
51287     MX-2014-110527
51288     MX-2014-114783
51289     CA-2014-156720
Name: order_id, Length: 51290, dtype: object

In [33]:
df.order_id[1:4]

1      IN-2011-47883
2       HU-2011-1220
3    IT-2011-3647632
Name: order_id, dtype: object

In [36]:
df.order_id[1:40:2]

1       IN-2011-47883
3     IT-2011-3647632
5       IN-2011-47883
7       IN-2011-79397
9        IZ-2011-4680
11      IN-2011-65159
13      IN-2011-33652
15     MX-2011-160234
17      ID-2011-80230
19      ID-2011-12596
21       IR-2011-7690
23       TZ-2011-7370
25      IN-2011-65159
27     MX-2011-111255
29     MX-2011-158771
31     MX-2011-111255
33     US-2011-136007
35     MX-2011-159373
37    IT-2011-2942451
39      IN-2011-33036
Name: order_id, dtype: object

In [ ]:
# creating new column

In [37]:
df['new_col'] = "Some_data"  

In [38]:
df

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... sub_category  \
0       Africa  ...      Storage   
1      Oceania  ...     Supplies   
2         EMEA  ...      Storage   
3        North  ...        Paper   
4      Oceania  ...  Furnishings   
...        ...  ...          ...   
51285     West  ...      Binders   
51286   Africa  ...      Binders   
51287  Central  ...       Labels   
51288    North  ...       Labels   
51289     West  ...    Fasteners   

                                          product_name sales quantity  \
0                                  Tenex Lockers, Blue   408        2   
1                             Acme Trimmer, High Speed   120        3   
2                              Tenex Box, Single Width    66        4   
3                          Enermax Note Cards, Premium    45        3   
4                           Eldon Light Bulb, Duo Pack   114        5   
...                                                ...   ...      ...   
51285  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl    14        2   
51286          Wilson Jones Hole Reinforcements, Clear     4        1   
51287           Hon Color Coded Labels, 5000 Label Set    26        3   
51288           Hon Legal Exhibit Labels, Alphabetical     7        1   
51289                              Bagged Rubber Bands     3        3   

      discount    profit  shipping_cost  order_priority  year    new_col  
0          0.0  106.1400          35.46          Medium  2011  Some_data  
1          0.1   36.0360           9.72          Medium  2011  Some_data  
2          0.0   29.6400           8.17            High  2011  Some_data  
3          0.5  -26.0550           4.82            High  2011  Some_data  
4          0.1   37.7700           4.70          Medium  2011  Some_data  
...        ...       ...            ...             ...   ...        ...  
51285      0.2    4.5188           0.89          Medium  2014  Some_data  
51286      0.0    0.4200           0.49          Medium  2014  Some_data  
51287      0.0   12.3600           0.35          Medium  2014  Some_data  
51288      0.0    0.5600           0.20          Medium  2014  Some_data  
51289      0.2   -0.6048           0.17          Medium  2014  Some_data  

[51290 rows x 22 columns]

In [39]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'new_col'],
      dtype='object')

In [41]:
df.order_id.isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
51285    False
51286    False
51287    False
51288    False
51289    False
Name: order_id, Length: 51290, dtype: bool

In [42]:
df.order_id.isnull().sum()

0

In [43]:
df2 = pd.read_csv('DressSales.csv')

In [44]:
df2

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0    1006032852       2114       2274      2491      2660      2727      2887   
1    1212192089        151        275       570       750       813      1066   
2    1190380701          6          7         7         7         8         8   
3     966005983       1005       1128      1326      1455      1507      1621   
4     876339541        996       1175      1304      1396      1432      1559   
..          ...        ...        ...       ...       ...       ...       ...   
495   713391965          0          0         0       560       554       544   
496   722565148          0          0         0       875       866       861   
497   532874347          0          0         0       734       728       726   
498   655464934          0          0         0       254       259       261   
499   919930954          0          0         0       538       545       558   

     10/9/2013 12/9/2013 14/9/2013  ... Unnamed: 26 Unnamed: 27 Unnamed: 28  \
0         2930      3119      3204  ...         NaN         NaN         NaN   
1         1164      1558      1756  ...         NaN         NaN         NaN   
2            9        10        10  ...         NaN         NaN         NaN   
3         1637      1723      1746  ...         NaN         NaN         NaN   
4         1570      1638      1655  ...         NaN         NaN         NaN   
..         ...       ...       ...  ...         ...         ...         ...   
495        537       525       519  ...         NaN         NaN         NaN   
496        854       850       844  ...         NaN         NaN         NaN   
497        715       694       690  ...         NaN         NaN         NaN   
498        263       268       270  ...         NaN         NaN         NaN   
499        563       578       585  ...         NaN         NaN         NaN   

    Unnamed: 29  Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  \
0           NaN          NaN          NaN          NaN          NaN   
1           NaN          NaN          NaN          NaN          NaN   
2           NaN          NaN          NaN          NaN          NaN   
3           NaN          NaN          NaN          NaN          NaN   
4           NaN          NaN          NaN          NaN          NaN   
..          ...          ...          ...          ...          ...   
495         NaN          NaN          NaN          NaN          NaN   
496         NaN          NaN          NaN          NaN          NaN   
497         NaN          NaN          NaN          NaN          NaN   
498         NaN          NaN          NaN          NaN          NaN   
499         NaN          NaN          NaN          NaN          NaN   

     Unnamed: 34  Unnamed: 35  
0            NaN          NaN  
1            NaN          NaN  
2            NaN          NaN  
3            NaN          NaN  
4            NaN          NaN  
..           ...          ...  
495          NaN          NaN  
496          NaN          NaN  
497          NaN          NaN  
498          NaN          NaN  
499          NaN          NaN  

[500 rows x 36 columns]

In [46]:
df2.isnull().sum()

Dress_ID         0
29/8/2013        0
31/8/2013        0
2/9/2013         0
4/9/2013         0
6/9/2013         0
8/9/2013         0
10/9/2013        0
12/9/2013        0
14/9/2013        0
16/9/2013        0
18/9/2013        0
20/9/2013        0
22/9/2013        0
24/9/2013        0
26/9/2013      237
28/9/2013        0
30/9/2013      273
2/10/2013      275
4/10/2013      274
6/10/2013        0
8/10/2010      271
10/10/2013     271
12/10/2013       0
Unnamed: 24    500
Unnamed: 25    500
Unnamed: 26    500
Unnamed: 27    500
Unnamed: 28    500
Unnamed: 29    500
Unnamed: 30    500
Unnamed: 31    500
Unnamed: 32    500
Unnamed: 33    500
Unnamed: 34    500
Unnamed: 35    500
dtype: int64

In [47]:
df2.columns

Index(['Dress_ID', '29/8/2013', '31/8/2013', '2/9/2013', '4/9/2013',
       '6/9/2013', '8/9/2013', '10/9/2013', '12/9/2013', '14/9/2013',
       '16/9/2013', '18/9/2013', '20/9/2013', '22/9/2013', '24/9/2013',
       '26/9/2013', '28/9/2013', '30/9/2013', '2/10/2013', '4/10/2013',
       '6/10/2013', '8/10/2010', '10/10/2013', '12/10/2013', 'Unnamed: 24',
       'Unnamed: 25', 'Unnamed: 26', 'Unnamed: 27', 'Unnamed: 28',
       'Unnamed: 29', 'Unnamed: 30', 'Unnamed: 31', 'Unnamed: 32',
       'Unnamed: 33', 'Unnamed: 34', 'Unnamed: 35'],
      dtype='object')

In [50]:
df2['Unnamed: 26'].isnull()

0      True
1      True
2      True
3      True
4      True
       ... 
495    True
496    True
497    True
498    True
499    True
Name: Unnamed: 26, Length: 500, dtype: bool

In [59]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ... sub_category  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...      Storage   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...     Supplies   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...      Storage   
3  Home Office        Stockholm     Sweden      EU    North  ...        Paper   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...  Furnishings   

                  product_name sales quantity discount   profit  \
0          Tenex Lockers, Blue   408        2      0.0  106.140   
1     Acme Trimmer, High Speed   120        3      0.1   36.036   
2      Tenex Box, Single Width    66        4      0.0   29.640   
3  Enermax Note Cards, Premium    45        3      0.5  -26.055   
4   Eldon Light Bulb, Duo Pack   114        5      0.1   37.770   

   shipping_cost  order_priority  year    new_col  
0          35.46          Medium  2011  Some_data  
1           9.72          Medium  2011  Some_data  
2           8.17            High  2011  Some_data  
3           4.82            High  2011  Some_data  
4           4.70          Medium  2011  Some_data  

[5 rows x 22 columns]

In [51]:
# Person name with maximum profit

In [55]:
df['profit'] == max(df['profit'])

0        False
1        False
2        False
3        False
4        False
         ...  
51285    False
51286    False
51287    False
51288    False
51289    False
Name: profit, Length: 51290, dtype: bool

In [56]:
df[df['profit'] == max(df['profit'])]['customer_name']

29530    Tamara Chand
Name: customer_name, dtype: object

In [57]:
# filter out all the records where country is 'Swedon'

In [60]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'new_col'],
      dtype='object')

In [63]:
df.country.unique()

array(['Algeria', 'Australia', 'Hungary', 'Sweden', 'Canada',
       'New Zealand', 'Iraq', 'Philippines', 'United Kingdom', 'Malaysia',
       'Guatemala', 'Iran', 'Thailand', 'Tanzania', 'Brazil', 'Mexico',
       'Cuba', 'France', 'United States', 'Japan', 'Sudan', 'Taiwan',
       'Indonesia', 'Vietnam', 'Angola', 'China', 'Mozambique', 'Lebanon',
       'Singapore', 'Netherlands', 'Nigeria', 'Egypt', 'Venezuela',
       'South Africa', 'Spain', 'India', 'Turkey', 'Austria', 'Italy',
       'Germany', 'Nicaragua', 'Dominican Republic', 'El Salvador',
       'Denmark', 'Saudi Arabia', 'Zambia', 'Myanmar (Burma)', 'Russia',
       'Mongolia', 'Belgium', 'Kenya', 'Colombia', 'Estonia',
       'Madagascar', 'Portugal', 'Morocco', 'Sierra Leone', 'Norway',
       'Central African Republic', 'Czech Republic', 'Benin',
       'Bangladesh', 'Panama', 'Chile', 'South Korea', 'Switzerland',
       'Moldova', 'Uganda', 'Zimbabwe', 'Niger', 'Senegal', 'Hong Kong',
       'Democratic Republic o

In [64]:
df[df['country']=='Sweden']

order_id  order_date   ship_date       ship_mode  \
3      IT-2011-3647632    1/1/2011    1/5/2011    Second Class   
168    IT-2011-5951216   1/13/2011   1/20/2011  Standard Class   
477    IT-2011-5645183    2/3/2011    2/5/2011    Second Class   
1342   IT-2011-5233011   3/31/2011    4/6/2011  Standard Class   
1554   IT-2011-2419493   4/12/2011   4/13/2011     First Class   
...                ...         ...         ...             ...   
50948  IT-2014-5217096  12/26/2014  12/31/2014  Standard Class   
50949  IT-2014-5217096  12/26/2014  12/31/2014  Standard Class   
50972  IT-2014-4046179  12/27/2014  12/30/2014    Second Class   
51086  IT-2014-2362249  12/29/2014    1/1/2015     First Class   
51121  IT-2014-2362249  12/29/2014    1/1/2015     First Class   

         customer_name      segment            state country market region  \
3         Eugene Moren  Home Office        Stockholm  Sweden     EU  North   
168     Stuart Calhoun     Consumer  V„stra G”taland  Sweden     EU  North   
477         Max Ludwig  Home Office          Uppsala  Sweden     EU  North   
1342      Todd Sumrall    Corporate  V„stra G”taland  Sweden     EU  North   
1554    Alice McCarthy    Corporate        Stockholm  Sweden     EU  North   
...                ...          ...              ...     ...    ...    ...   
50948     Trudy Glocke     Consumer        Stockholm  Sweden     EU  North   
50949     Trudy Glocke     Consumer        Stockholm  Sweden     EU  North   
50972     Jeremy Farry     Consumer        Stockholm  Sweden     EU  North   
51086  Mitch Gastineau    Corporate        Stockholm  Sweden     EU  North   
51121  Mitch Gastineau    Corporate        Stockholm  Sweden     EU  North   

       ... sub_category                         product_name  sales quantity  \
3      ...        Paper          Enermax Note Cards, Premium     45        3   
168    ...          Art            Sanford Pens, Fluorescent     31        5   
477    ...          Art  Boston Pencil Sharpener, Easy-Erase     91        6   
1342   ...  Furnishings        Advantus Photo Frame, Durable     40        2   
1554   ...       Phones                  Nokia Headset, VoIP     73        2   
...    ...          ...                                  ...    ...      ...   
50948  ...  Accessories        Belkin Flash Drive, Bluetooth     21        1   
50949  ...  Furnishings        Advantus Photo Frame, Durable     20        1   
50972  ...   Appliances       KitchenAid Refrigerator, Black  1,582        6   
51086  ...          Art                BIC Pens, Fluorescent     31        4   
51121  ...          Art           Sanford Highlighters, Blue      8        1   

      discount   profit  shipping_cost  order_priority  year    new_col  
3          0.5  -26.055           4.82            High  2011  Some_data  
168        0.5  -13.650           2.16          Medium  2011  Some_data  
477        0.5  -61.650           6.64          Medium  2011  Some_data  
1342       0.6  -44.052           2.78          Medium  2011  Some_data  
1554       0.5  -11.760          22.18            High  2011  Some_data  
...        ...      ...            ...             ...   ...        ...  
50948      0.5   -5.460           0.98          Medium  2014  Some_data  
50949      0.6  -22.026           0.98          Medium  2014  Some_data  
50972      0.5 -443.070          55.39          Medium  2014  Some_data  
51086      0.5  -10.680           8.44            High  2014  Some_data  
51121      0.5   -6.165           2.00            High  2014  Some_data  

[203 rows x 22 columns]

In [65]:
df[df['shipping_cost'] > 80]

order_id  order_date ship_date       ship_mode  \
7       IN-2011-79397    1/3/2011  1/3/2011        Same Day   
8       ID-2011-80230    1/3/2011  1/9/2011  Standard Class   
9        IZ-2011-4680    1/3/2011  1/7/2011  Standard Class   
27     MX-2011-111255    1/4/2011  1/9/2011    Second Class   
43     MX-2011-109267    1/5/2011  1/9/2011  Standard Class   
...               ...         ...       ...             ...   
51154   IN-2014-49206  12/30/2014  1/1/2015    Second Class   
51228  MX-2014-116267  12/31/2014  1/3/2015    Second Class   
51229   IN-2014-43550  12/31/2014  1/1/2015     First Class   
51230    RS-2014-1460  12/31/2014  1/2/2015    Second Class   
51231   IN-2014-30390  12/31/2014  1/3/2015     First Class   

             customer_name    segment            state      country market  \
7              Kean Nguyen  Corporate  New South Wales    Australia   APAC   
8             Ken Lonsdale   Consumer         Auckland  New Zealand   APAC   
9         Lindsay Williams  Corporate           Ninawa         Iraq   EMEA   
27       Russell Applegate   Consumer           Parana       Brazil  LATAM   
43       Jennifer Halladay   Consumer         Veracruz       Mexico  LATAM   
...                    ...        ...              ...          ...    ...   
51154          Toby Ritter   Consumer       Jawa Barat    Indonesia   APAC   
51228           Erica Bern  Corporate        S?o Paulo       Brazil  LATAM   
51229  Marina Lichtenstein  Corporate          Jakarta    Indonesia   APAC   
51230         Peter Bhler   Consumer    Bashkortostan       Russia   EMEA   
51231     Justin Deggeller  Corporate          Bangkok     Thailand   APAC   

               region  ... sub_category  \
7             Oceania  ...   Appliances   
8             Oceania  ...      Copiers   
9                EMEA  ...       Chairs   
27              South  ...    Bookcases   
43              North  ...   Appliances   
...               ...  ...          ...   
51154  Southeast Asia  ...      Storage   
51228           South  ...      Copiers   
51229  Southeast Asia  ...    Bookcases   
51230            EMEA  ...      Copiers   
51231  Southeast Asia  ...       Tables   

                                  product_name  sales quantity discount  \
7                      Hoover Microwave, White    276        1    0.100   
8                  Hewlett Wireless Fax, Laser    912        4    0.400   
9             Novimex Swivel Stool, Set of Two    667        4    0.000   
27                Dania Classic Bookcase, Pine  1,648        6    0.000   
43                         Hoover Stove, Black  3,029        8    0.000   
...                                        ...    ...      ...      ...   
51154              Rogers Shelving, Wire Frame    457        9    0.170   
51228                Canon Wireless Fax, Color  1,264        5    0.002   
51229           Safco Library with Doors, Pine  1,091        3    0.070   
51230           Hewlett Personal Copier, Laser    866        6    0.000   
51231  Lesro Training Table, Adjustable Height  1,049        9    0.570   

          profit  shipping_cost  order_priority  year    new_col  
7       110.4120         125.32        Critical  2011  Some_data  
8      -319.4640         107.10             Low  2011  Some_data  
9       253.3200          81.26            High  2011  Some_data  
27      609.8400         109.13          Medium  2011  Some_data  
43      999.3600         191.20          Medium  2011  Some_data  
...          ...            ...             ...   ...        ...  
51154    65.9340         101.88            High  2014  Some_data  
51228   301.4660         253.25            High  2014  Some_data  
51229    46.9206         243.11            High  2014  Some_data  
51230    51.8400         138.18            High  2014  Some_data  
51231 -1195.2387          86.86            High  2014  Some_data  

[3984 rows x 22 columns]

In [66]:
df[df['shipping_cost'] > 80]['country']

7          Australia
8        New Zealand
9               Iraq
27            Brazil
43            Mexico
            ...     
51154      Indonesia
51228         Brazil
51229      Indonesia
51230         Russia
51231       Thailand
Name: country, Length: 3984, dtype: object

In [72]:
df[(df['shipping_cost'] > 100) & (df['profit'] < 10)]

order_id  order_date   ship_date       ship_mode  \
8        ID-2011-80230    1/3/2011    1/9/2011  Standard Class   
257     MX-2011-146584   1/21/2011   1/27/2011  Standard Class   
451    ES-2011-1640672    2/2/2011    2/5/2011     First Class   
625      ID-2011-47575   2/15/2011   2/21/2011  Standard Class   
626      ID-2011-47575   2/15/2011   2/21/2011  Standard Class   
...                ...         ...         ...             ...   
50781   CA-2014-147956  12/25/2014    1/1/2015  Standard Class   
50969   US-2014-120166  12/27/2014  12/31/2014  Standard Class   
50970  ES-2014-4296634  12/27/2014    1/2/2015  Standard Class   
51034    ID-2014-68183  12/29/2014  12/31/2014     First Class   
51153    ID-2014-51509  12/30/2014    1/1/2015    Second Class   

              customer_name    segment                         state  \
8              Ken Lonsdale   Consumer                      Auckland   
257            Hunter Lopez   Consumer                       Managua   
451              Ken Heidel  Corporate                        Saxony   
625    Christopher Martinez   Consumer              National Capital   
626    Christopher Martinez   Consumer              National Capital   
...                     ...        ...                           ...   
50781            Alan Hwang   Consumer                    Washington   
50969          Frank Hawley  Corporate               Valle del Cauca   
50970           Harry Marie  Corporate                      Brittany   
51034             Adam Hart  Corporate                      Victoria   
51153          Naresj Patel   Consumer  Australian Capital Territory   

             country market          region  ... sub_category  \
8        New Zealand   APAC         Oceania  ...      Copiers   
257        Nicaragua  LATAM         Central  ...       Tables   
451          Germany     EU         Central  ...       Phones   
625      Philippines   APAC  Southeast Asia  ...       Phones   
626      Philippines   APAC  Southeast Asia  ...      Copiers   
...              ...    ...             ...  ...          ...   
50781  United States     US            West  ...      Storage   
50969       Colombia  LATAM           South  ...       Tables   
50970         France     EU         Central  ...   Appliances   
51034      Australia   APAC         Oceania  ...   Appliances   
51153      Australia   APAC         Oceania  ...       Chairs   

                                            product_name  sales quantity  \
8                            Hewlett Wireless Fax, Laser    912        4   
257         Chromcraft Conference Table, Fully Assembled  2,797        6   
451                    Motorola Signal Booster, Cordless    584        8   
625                 Nokia Signal Booster, with Caller ID    921        9   
626                      Brother Fax Machine, High-Speed  1,235        6   
...                                                  ...    ...      ...   
50781  Office Impressions Heavy Duty Welded Shelving ...  1,004        6   
50969               Barricks Training Table, Rectangular    842        5   
50970                    Hamilton Beach Microwave, White  1,010        4   
51034                                  Hoover Stove, Red  1,535        3   
51153             Harbour Creations Rocking Chair, Black    434        5   

      discount   profit  shipping_cost  order_priority  year    new_col  
8         0.40 -319.464         107.10             Low  2011  Some_data  
257       0.20   -0.072         320.38             Low  2011  Some_data  
451       0.50  -58.560         109.14            High  2011  Some_data  
625       0.25 -172.125         128.01             Low  2011  Some_data  
626       0.35 -456.174         118.63             Low  2011  Some_data  
...        ...      ...            ...             ...   ...        ...  
50781     0.00    0.000         121.83             Low  2014  Some_data  
50969     0.20 -200.080         120.11            High  2014  Some_dat

In [75]:
df[(df['shipping_cost'] > 100) & (df['profit'] < 10)]['country']

8          New Zealand
257          Nicaragua
451            Germany
625        Philippines
626        Philippines
             ...      
50781    United States
50969         Colombia
50970           France
51034        Australia
51153        Australia
Name: country, Length: 604, dtype: object

In [76]:
df.dtypes

order_id           object
order_date         object
ship_date          object
ship_mode          object
customer_name      object
segment            object
state              object
country            object
market             object
region             object
product_id         object
category           object
sub_category       object
product_name       object
sales              object
quantity            int64
discount          float64
profit            float64
shipping_cost     float64
order_priority     object
year                int64
new_col            object
dtype: object

In [85]:
df[['customer_name','profit']].max()

customer_name    Zuschuss Donatelli
profit                     8399.976
dtype: object

In [77]:
# date-time operations

In [80]:
df['new_order_date'] = pd.to_datetime(df['order_date'])

In [81]:
df.dtypes

order_id                  object
order_date                object
ship_date                 object
ship_mode                 object
customer_name             object
segment                   object
state                     object
country                   object
market                    object
region                    object
product_id                object
category                  object
sub_category              object
product_name              object
sales                     object
quantity                   int64
discount                 float64
profit                   float64
shipping_cost            float64
order_priority            object
year                       int64
new_col                   object
new_order_date    datetime64[ns]
dtype: object

In [82]:
df['year'] = df['new_order_date'].dt.year

In [83]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ...  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...   
3  Home Office        Stockholm     Sweden      EU    North  ...   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...   

                  product_name sales quantity discount   profit  \
0          Tenex Lockers, Blue   408        2      0.0  106.140   
1     Acme Trimmer, High Speed   120        3      0.1   36.036   
2      Tenex Box, Single Width    66        4      0.0   29.640   
3  Enermax Note Cards, Premium    45        3      0.5  -26.055   
4   Eldon Light Bulb, Duo Pack   114        5      0.1   37.770   

   shipping_cost  order_priority  year    new_col new_order_date  
0          35.46          Medium  2011  Some_data     2011-01-01  
1           9.72          Medium  2011  Some_data     2011-01-01  
2           8.17            High  2011  Some_data     2011-01-01  
3           4.82            High  2011  Some_data     2011-01-01  
4           4.70          Medium  2011  Some_data     2011-01-01  

[5 rows x 23 columns]

In [84]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'new_col', 'new_order_date'],
      dtype='object')

In [86]:
df['order_month'] = df['new_order_date'].dt.month

In [87]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ... sales  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...   408   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...   120   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...    66   
3  Home Office        Stockholm     Sweden      EU    North  ...    45   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...   114   

  quantity discount   profit shipping_cost  order_priority  year    new_col  \
0        2      0.0  106.140         35.46          Medium  2011  Some_data   
1        3      0.1   36.036          9.72          Medium  2011  Some_data   
2        4      0.0   29.640          8.17            High  2011  Some_data   
3        3      0.5  -26.055          4.82            High  2011  Some_data   
4        5      0.1   37.770          4.70          Medium  2011  Some_data   

   new_order_date order_month  
0      2011-01-01           1  
1      2011-01-01           1  
2      2011-01-01           1  
3      2011-01-01           1  
4      2011-01-01           1  

[5 rows x 24 columns]

In [88]:
df['order_month'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=int64)

In [91]:
df['order_weak'] = df['new_order_date'].dt.week

<ipython-input-91-8d1feb1aa19c>:1: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  df['order_weak'] = df['new_order_date'].dt.week


In [93]:
df[['order_month','order_weak']]  # shows number of week left

order_month  order_weak
0                1          52
1                1          52
2                1          52
3                1          52
4                1          52
...            ...         ...
51285           12           1
51286           12           1
51287           12           1
51288           12           1
51289           12           1

[51290 rows x 2 columns]

In [94]:
# showing data from 2014

In [97]:
df[df['year']==2014].head()

order_id order_date ship_date       ship_mode  customer_name  \
33759  ES-2014-2774938   1/1/2014  1/3/2014    Second Class    Fred Harton   
33760    IN-2014-47134   1/1/2014  1/5/2014  Standard Class    Kean Nguyen   
33761  IT-2014-4810734   1/1/2014  1/3/2014    Second Class   Alan Shonely   
33762   US-2014-105830   1/1/2014  1/3/2014     First Class   Duane Benoit   
33763   CA-2014-160395   1/1/2014  1/7/2014  Standard Class  Kelly Lampkin   

         segment          state        country market   region  ... quantity  \
33759   Consumer  Ile-de-France         France     EU  Central  ...        3   
33760  Corporate       Victoria      Australia   APAC  Oceania  ...        3   
33761   Consumer  Ile-de-France         France     EU  Central  ...        6   
33762   Consumer           Ohio  United States     US     East  ...        4   
33763  Corporate         Nevada  United States     US     West  ...        7   

      discount    profit shipping_cost order_priority  year    new_col  \
33759     0.15 -111.6225        104.69           High  2014  Some_data   
33760     0.10   46.9980         46.20         Medium  2014  Some_data   
33761     0.15  -43.1370         37.85       Critical  2014  Some_data   
33762     0.20  -35.2152         30.66           High  2014  Some_data   
33763     0.20   53.9217         26.90            Low  2014  Some_data   

       new_order_date  order_month order_weak  
33759      2014-01-01            1          1  
33760      2014-01-01            1          1  
33761      2014-01-01            1          1  
33762      2014-01-01            1          1  
33763      2014-01-01            1          1  

[5 rows x 25 columns]

In [101]:
#df[df['order_month']==max(df['quantity'].sum())]

In [99]:
df['order_month'].value_counts()

11    6342
12    6302
9     5975
6     5331
8     5078
10    4490
5     3747
7     3166
4     3057
3     3035
1     2599
2     2168
Name: order_month, dtype: int64

In [102]:
#df[max(df['order_month'].value_counts())]

In [103]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'new_col', 'new_order_date', 'order_month', 'order_weak'],
      dtype='object')

In [104]:
df['cost_to_company'] = df['discount'] + df['shipping_cost']

In [106]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ... discount  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...      0.0   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...      0.1   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...      0.0   
3  Home Office        Stockholm     Sweden      EU    North  ...      0.5   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...      0.1   

    profit shipping_cost order_priority  year    new_col  new_order_date  \
0  106.140         35.46         Medium  2011  Some_data      2011-01-01   
1   36.036          9.72         Medium  2011  Some_data      2011-01-01   
2   29.640          8.17           High  2011  Some_data      2011-01-01   
3  -26.055          4.82           High  2011  Some_data      2011-01-01   
4   37.770          4.70         Medium  2011  Some_data      2011-01-01   

   order_month  order_weak cost_to_company  
0            1          52           35.46  
1            1          52            9.82  
2            1          52            8.17  
3            1          52            5.32  
4            1          52            4.80  

[5 rows x 26 columns]

In [110]:
df[df['cost_to_company'] == max(df['cost_to_company'])]['product_name']

13891    Plantronics CS510 - Over-the-Head monaural Wir...
Name: product_name, dtype: object

In [111]:
df.drop('year', axis=1)

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... quantity discount    profit shipping_cost order_priority  \
0       Africa  ...        2      0.0  106.1400         35.46         Medium   
1      Oceania  ...        3      0.1   36.0360          9.72         Medium   
2         EMEA  ...        4      0.0   29.6400          8.17           High   
3        North  ...        3      0.5  -26.0550          4.82           High   
4      Oceania  ...        5      0.1   37.7700          4.70         Medium   
...        ...  ...      ...      ...       ...           ...            ...   
51285     West  ...        2      0.2    4.5188          0.89         Medium   
51286   Africa  ...        1      0.0    0.4200          0.49         Medium   
51287  Central  ...        3      0.0   12.3600          0.35         Medium   
51288    North  ...        1      0.0    0.5600          0.20         Medium   
51289     West  ...        3      0.2   -0.6048          0.17         Medium   

         new_col  new_order_date  order_month  order_weak cost_to_company  
0      Some_data      2011-01-01            1          52           35.46  
1      Some_data      2011-01-01            1          52            9.82  
2      Some_data      2011-01-01            1          52            8.17  
3      Some_data      2011-01-01            1          52            5.32  
4      Some_data      2011-01-01            1          52            4.80  
...          ...             ...          ...         ...             ...  
51285  Some_data      2014-12-31           12           1            1.09  
51286  Some_data      2014-12-31           12           1            0.49  
51287  Some_data      2014-12-31           12           1            0.35  
51288  Some_data      2014-12-31           12           1            0.20  
51289  Some_data      2014-12-31           12           1            0.37  

[51290 rows x 25 columns]

In [112]:
df.drop(['year','new_col'], axis=1)

order_id  order_date ship_date       ship_mode  \
0         AG-2011-2040    1/1/2011  1/6/2011  Standard Class   
1        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
2         HU-2011-1220    1/1/2011  1/5/2011    Second Class   
3      IT-2011-3647632    1/1/2011  1/5/2011    Second Class   
4        IN-2011-47883    1/1/2011  1/8/2011  Standard Class   
...                ...         ...       ...             ...   
51285   CA-2014-115427  12/31/2014  1/4/2015  Standard Class   
51286     MO-2014-2560  12/31/2014  1/5/2015  Standard Class   
51287   MX-2014-110527  12/31/2014  1/2/2015    Second Class   
51288   MX-2014-114783  12/31/2014  1/6/2015  Standard Class   
51289   CA-2014-156720  12/31/2014  1/4/2015  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draƒ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... sales quantity discount    profit shipping_cost  \
0       Africa  ...   408        2      0.0  106.1400         35.46   
1      Oceania  ...   120        3      0.1   36.0360          9.72   
2         EMEA  ...    66        4      0.0   29.6400          8.17   
3        North  ...    45        3      0.5  -26.0550          4.82   
4      Oceania  ...   114        5      0.1   37.7700          4.70   
...        ...  ...   ...      ...      ...       ...           ...   
51285     West  ...    14        2      0.2    4.5188          0.89   
51286   Africa  ...     4        1      0.0    0.4200          0.49   
51287  Central  ...    26        3      0.0   12.3600          0.35   
51288    North  ...     7        1      0.0    0.5600          0.20   
51289     West  ...     3        3      0.2   -0.6048          0.17   

       order_priority  new_order_date  order_month  order_weak cost_to_company  
0              Medium      2011-01-01            1          52           35.46  
1              Medium      2011-01-01            1          52            9.82  
2                High      2011-01-01            1          52            8.17  
3                High      2011-01-01            1          52            5.32  
4              Medium      2011-01-01            1          52            4.80  
...               ...             ...          ...         ...             ...  
51285          Medium      2014-12-31           12           1            1.09  
51286          Medium      2014-12-31           12           1            0.49  
51287          Medium      2014-12-31           12           1            0.35  
51288          Medium      2014-12-31           12           1            0.20  
51289          Medium      2014-12-31           12           1            0.37  

[51290 rows x 24 columns]

In [113]:
# making change permanent (two ways)

In [114]:
df = df.drop(['year','new_col'], axis=1)

In [115]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ... sales  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...   408   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...   120   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...    66   
3  Home Office        Stockholm     Sweden      EU    North  ...    45   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...   114   

  quantity discount   profit shipping_cost  order_priority  new_order_date  \
0        2      0.0  106.140         35.46          Medium      2011-01-01   
1        3      0.1   36.036          9.72          Medium      2011-01-01   
2        4      0.0   29.640          8.17            High      2011-01-01   
3        3      0.5  -26.055          4.82            High      2011-01-01   
4        5      0.1   37.770          4.70          Medium      2011-01-01   

   order_month  order_weak cost_to_company  
0            1          52           35.46  
1            1          52            9.82  
2            1          52            8.17  
3            1          52            5.32  
4            1          52            4.80  

[5 rows x 24 columns]

In [116]:
df.drop('order_weak',axis=1, inplace=True)

In [117]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ...  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...   
3  Home Office        Stockholm     Sweden      EU    North  ...   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...   

                  product_name sales quantity discount   profit  \
0          Tenex Lockers, Blue   408        2      0.0  106.140   
1     Acme Trimmer, High Speed   120        3      0.1   36.036   
2      Tenex Box, Single Width    66        4      0.0   29.640   
3  Enermax Note Cards, Premium    45        3      0.5  -26.055   
4   Eldon Light Bulb, Duo Pack   114        5      0.1   37.770   

   shipping_cost  order_priority  new_order_date  order_month cost_to_company  
0          35.46          Medium      2011-01-01            1           35.46  
1           9.72          Medium      2011-01-01            1            9.82  
2           8.17            High      2011-01-01            1            8.17  
3           4.82            High      2011-01-01            1            5.32  
4           4.70          Medium      2011-01-01            1            4.80  

[5 rows x 23 columns]

In [119]:
# row operations (loc, iloc)

In [124]:
df.loc[[1,3]]

order_id order_date ship_date       ship_mode customer_name  \
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class   Joseph Holt   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class  Eugene Moren   

       segment            state    country market   region  ...  \
1     Consumer  New South Wales  Australia   APAC  Oceania  ...   
3  Home Office        Stockholm     Sweden     EU    North  ...   

                  product_name sales quantity discount  profit  shipping_cost  \
1     Acme Trimmer, High Speed   120        3      0.1  36.036           9.72   
3  Enermax Note Cards, Premium    45        3      0.5 -26.055           4.82   

   order_priority  new_order_date  order_month cost_to_company  
1          Medium      2011-01-01            1            9.82  
3            High      2011-01-01            1            5.32  

[2 rows x 23 columns]

In [126]:
df.loc[1:10]

order_id order_date ship_date       ship_mode     customer_name  \
1     IN-2011-47883   1/1/2011  1/8/2011  Standard Class       Joseph Holt   
2      HU-2011-1220   1/1/2011  1/5/2011    Second Class     Annie Thurman   
3   IT-2011-3647632   1/1/2011  1/5/2011    Second Class      Eugene Moren   
4     IN-2011-47883   1/1/2011  1/8/2011  Standard Class       Joseph Holt   
5     IN-2011-47883   1/1/2011  1/8/2011  Standard Class       Joseph Holt   
6      CA-2011-1510   1/2/2011  1/6/2011  Standard Class   Magdelene Morse   
7     IN-2011-79397   1/3/2011  1/3/2011        Same Day       Kean Nguyen   
8     ID-2011-80230   1/3/2011  1/9/2011  Standard Class      Ken Lonsdale   
9      IZ-2011-4680   1/3/2011  1/7/2011  Standard Class  Lindsay Williams   
10    IN-2011-65159   1/3/2011  1/7/2011    Second Class      Larry Blacks   

        segment             state      country  market          region  ...  \
1      Consumer   New South Wales    Australia    APAC         Oceania  ...   
2      Consumer          Budapest      Hungary    EMEA            EMEA  ...   
3   Home Office         Stockholm       Sweden      EU           North  ...   
4      Consumer   New South Wales    Australia    APAC         Oceania  ...   
5      Consumer   New South Wales    Australia    APAC         Oceania  ...   
6      Consumer           Ontario       Canada  Canada          Canada  ...   
7     Corporate   New South Wales    Australia    APAC         Oceania  ...   
8      Consumer          Auckland  New Zealand    APAC         Oceania  ...   
9     Corporate            Ninawa         Iraq    EMEA            EMEA  ...   
10     Consumer  National Capital  Philippines    APAC  Southeast Asia  ...   

                               product_name sales quantity discount    profit  \
1                  Acme Trimmer, High Speed   120        3     0.10   36.0360   
2                   Tenex Box, Single Width    66        4     0.00   29.6400   
3               Enermax Note Cards, Premium    45        3     0.50  -26.0550   
4                Eldon Light Bulb, Duo Pack   114        5     0.10   37.7700   
5   Eaton Computer Printout Paper, 8.5 x 11    55        2     0.10   15.3420   
6                  Okidata Inkjet, Wireless   314        1     0.00    3.1200   
7                   Hoover Microwave, White   276        1     0.10  110.4120   
8               Hewlett Wireless Fax, Laser   912        4     0.40 -319.4640   
9          Novimex Swivel Stool, Set of Two   667        4     0.00  253.3200   
10                Tenex Lockers, Industrial   338        3     0.45 -122.8005   

    shipping_cost  order_priority  new_order_date  order_month cost_to_company  
1            9.72          Medium      2011-01-01            1            9.82  
2            8.17            High      2011-01-01            1            8.17  
3            4.82            High      2011-01-01            1            5.32  
4            4.70          Medium      2011-01-01            1            4.80  
5            1.80          Medium      2011-01-01            1            1.90  
6           24.10          Medium      2011-01-02            1           24.10  
7          125.32        Critical      2011-01-03            1          125.42  
8          107.10             Low      2011-01-03            1          107.50  
9           81.26            High      2011-01-03            1           81.26  
10          33.75            High      2011-01-03            1           34.20  

[10 rows x 23 columns]

In [132]:
df.loc[0:4,['order_id','order_date', 'ship_date', 'ship_mode']]

order_id order_date ship_date       ship_mode
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class

In [131]:
df.iloc[0:3,0:3]

order_id order_date ship_date
0   AG-2011-2040   1/1/2011  1/6/2011
1  IN-2011-47883   1/1/2011  1/8/2011
2   HU-2011-1220   1/1/2011  1/5/2011

In [134]:
df.loc[['order_id':'ship_mode']]

SyntaxError: invalid syntax (<ipython-input-134-7389f6aab40f>, line 1)

# loc takes named index 

# iloc understand default indexes, internal indexes (used by system)

In [136]:
df.iloc[2:5,3:6]

ship_mode  customer_name      segment
2    Second Class  Annie Thurman     Consumer
3    Second Class   Eugene Moren  Home Office
4  Standard Class    Joseph Holt     Consumer

In [138]:
df2.head()

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0  1006032852       2114       2274      2491      2660      2727      2887   
1  1212192089        151        275       570       750       813      1066   
2  1190380701          6          7         7         7         8         8   
3   966005983       1005       1128      1326      1455      1507      1621   
4   876339541        996       1175      1304      1396      1432      1559   

   10/9/2013 12/9/2013 14/9/2013  ... Unnamed: 26 Unnamed: 27 Unnamed: 28  \
0       2930      3119      3204  ...         NaN         NaN         NaN   
1       1164      1558      1756  ...         NaN         NaN         NaN   
2          9        10        10  ...         NaN         NaN         NaN   
3       1637      1723      1746  ...         NaN         NaN         NaN   
4       1570      1638      1655  ...         NaN         NaN         NaN   

  Unnamed: 29  Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 34  Unnamed: 35  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN  

[5 rows x 36 columns]

In [139]:
df2.dropna()  # dropped all the columns

Empty DataFrame
Columns: [Dress_ID, 29/8/2013, 31/8/2013, 2/9/2013, 4/9/2013, 6/9/2013, 8/9/2013, 10/9/2013, 12/9/2013, 14/9/2013, 16/9/2013, 18/9/2013, 20/9/2013, 22/9/2013, 24/9/2013, 26/9/2013, 28/9/2013, 30/9/2013, 2/10/2013, 4/10/2013, 6/10/2013, 8/10/2010, 10/10/2013, 12/10/2013, Unnamed: 24, Unnamed: 25, Unnamed: 26, Unnamed: 27, Unnamed: 28, Unnamed: 29, Unnamed: 30, Unnamed: 31, Unnamed: 32, Unnamed: 33, Unnamed: 34, Unnamed: 35]
Index: []

[0 rows x 36 columns]

In [141]:
df2.dropna(axis=1).head()

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0  1006032852       2114       2274      2491      2660      2727      2887   
1  1212192089        151        275       570       750       813      1066   
2  1190380701          6          7         7         7         8         8   
3   966005983       1005       1128      1326      1455      1507      1621   
4   876339541        996       1175      1304      1396      1432      1559   

   10/9/2013 12/9/2013 14/9/2013 16/9/2013 18/9/2013 20/9/2013 22/9/2013  \
0       2930      3119      3204      3277      3321      3386      3479   
1       1164      1558      1756      1878      1985      2106      2454   
2          9        10        10        10        10        10        11   
3       1637      1723      1746      1783      1796      1812      1845   
4       1570      1638      1655      1681      1743      1824      1919   

   24/9/2013  28/9/2013  6/10/2013  12/10/2013  
0       3554       3706       3897        4048  
1       2710       3258       3911        4277  
2         11         11         11          11  
3       1878       1914       1952        1963  
4       2032       2252       2544        2736

In [149]:
df2.dropna(thresh=5) # atleast 5 Nan records

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0    1006032852       2114       2274      2491      2660      2727      2887   
1    1212192089        151        275       570       750       813      1066   
2    1190380701          6          7         7         7         8         8   
3     966005983       1005       1128      1326      1455      1507      1621   
4     876339541        996       1175      1304      1396      1432      1559   
..          ...        ...        ...       ...       ...       ...       ...   
495   713391965          0          0         0       560       554       544   
496   722565148          0          0         0       875       866       861   
497   532874347          0          0         0       734       728       726   
498   655464934          0          0         0       254       259       261   
499   919930954          0          0         0       538       545       558   

     10/9/2013 12/9/2013 14/9/2013  ... Unnamed: 26 Unnamed: 27 Unnamed: 28  \
0         2930      3119      3204  ...         NaN         NaN         NaN   
1         1164      1558      1756  ...         NaN         NaN         NaN   
2            9        10        10  ...         NaN         NaN         NaN   
3         1637      1723      1746  ...         NaN         NaN         NaN   
4         1570      1638      1655  ...         NaN         NaN         NaN   
..         ...       ...       ...  ...         ...         ...         ...   
495        537       525       519  ...         NaN         NaN         NaN   
496        854       850       844  ...         NaN         NaN         NaN   
497        715       694       690  ...         NaN         NaN         NaN   
498        263       268       270  ...         NaN         NaN         NaN   
499        563       578       585  ...         NaN         NaN         NaN   

    Unnamed: 29  Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  \
0           NaN          NaN          NaN          NaN          NaN   
1           NaN          NaN          NaN          NaN          NaN   
2           NaN          NaN          NaN          NaN          NaN   
3           NaN          NaN          NaN          NaN          NaN   
4           NaN          NaN          NaN          NaN          NaN   
..          ...          ...          ...          ...          ...   
495         NaN          NaN          NaN          NaN          NaN   
496         NaN          NaN          NaN          NaN          NaN   
497         NaN          NaN          NaN          NaN          NaN   
498         NaN          NaN          NaN          NaN          NaN   
499         NaN          NaN          NaN          NaN          NaN   

     Unnamed: 34  Unnamed: 35  
0            NaN          NaN  
1            NaN          NaN  
2            NaN          NaN  
3            NaN          NaN  
4            NaN          NaN  
..           ...          ...  
495          NaN          NaN  
496          NaN          NaN  
497          NaN          NaN  
498          NaN          NaN  
499          NaN          NaN  

[500 rows x 36 columns]

In [148]:
df2.dropna(thresh=3)

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0    1006032852       2114       2274      2491      2660      2727      2887   
1    1212192089        151        275       570       750       813      1066   
2    1190380701          6          7         7         7         8         8   
3     966005983       1005       1128      1326      1455      1507      1621   
4     876339541        996       1175      1304      1396      1432      1559   
..          ...        ...        ...       ...       ...       ...       ...   
495   713391965          0          0         0       560       554       544   
496   722565148          0          0         0       875       866       861   
497   532874347          0          0         0       734       728       726   
498   655464934          0          0         0       254       259       261   
499   919930954          0          0         0       538       545       558   

     10/9/2013 12/9/2013 14/9/2013  ... Unnamed: 26 Unnamed: 27 Unnamed: 28  \
0         2930      3119      3204  ...         NaN         NaN         NaN   
1         1164      1558      1756  ...         NaN         NaN         NaN   
2            9        10        10  ...         NaN         NaN         NaN   
3         1637      1723      1746  ...         NaN         NaN         NaN   
4         1570      1638      1655  ...         NaN         NaN         NaN   
..         ...       ...       ...  ...         ...         ...         ...   
495        537       525       519  ...         NaN         NaN         NaN   
496        854       850       844  ...         NaN         NaN         NaN   
497        715       694       690  ...         NaN         NaN         NaN   
498        263       268       270  ...         NaN         NaN         NaN   
499        563       578       585  ...         NaN         NaN         NaN   

    Unnamed: 29  Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  \
0           NaN          NaN          NaN          NaN          NaN   
1           NaN          NaN          NaN          NaN          NaN   
2           NaN          NaN          NaN          NaN          NaN   
3           NaN          NaN          NaN          NaN          NaN   
4           NaN          NaN          NaN          NaN          NaN   
..          ...          ...          ...          ...          ...   
495         NaN          NaN          NaN          NaN          NaN   
496         NaN          NaN          NaN          NaN          NaN   
497         NaN          NaN          NaN          NaN          NaN   
498         NaN          NaN          NaN          NaN          NaN   
499         NaN          NaN          NaN          NaN          NaN   

     Unnamed: 34  Unnamed: 35  
0            NaN          NaN  
1            NaN          NaN  
2            NaN          NaN  
3            NaN          NaN  
4            NaN          NaN  
..           ...          ...  
495          NaN          NaN  
496          NaN          NaN  
497          NaN          NaN  
498          NaN          NaN  
499          NaN          NaN  

[500 rows x 36 columns]

In [147]:
df2.fillna(value=4)

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0    1006032852       2114       2274      2491      2660      2727      2887   
1    1212192089        151        275       570       750       813      1066   
2    1190380701          6          7         7         7         8         8   
3     966005983       1005       1128      1326      1455      1507      1621   
4     876339541        996       1175      1304      1396      1432      1559   
..          ...        ...        ...       ...       ...       ...       ...   
495   713391965          0          0         0       560       554       544   
496   722565148          0          0         0       875       866       861   
497   532874347          0          0         0       734       728       726   
498   655464934          0          0         0       254       259       261   
499   919930954          0          0         0       538       545       558   

     10/9/2013 12/9/2013 14/9/2013  ... Unnamed: 26 Unnamed: 27 Unnamed: 28  \
0         2930      3119      3204  ...         4.0         4.0         4.0   
1         1164      1558      1756  ...         4.0         4.0         4.0   
2            9        10        10  ...         4.0         4.0         4.0   
3         1637      1723      1746  ...         4.0         4.0         4.0   
4         1570      1638      1655  ...         4.0         4.0         4.0   
..         ...       ...       ...  ...         ...         ...         ...   
495        537       525       519  ...         4.0         4.0         4.0   
496        854       850       844  ...         4.0         4.0         4.0   
497        715       694       690  ...         4.0         4.0         4.0   
498        263       268       270  ...         4.0         4.0         4.0   
499        563       578       585  ...         4.0         4.0         4.0   

    Unnamed: 29  Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  \
0           4.0          4.0          4.0          4.0          4.0   
1           4.0          4.0          4.0          4.0          4.0   
2           4.0          4.0          4.0          4.0          4.0   
3           4.0          4.0          4.0          4.0          4.0   
4           4.0          4.0          4.0          4.0          4.0   
..          ...          ...          ...          ...          ...   
495         4.0          4.0          4.0          4.0          4.0   
496         4.0          4.0          4.0          4.0          4.0   
497         4.0          4.0          4.0          4.0          4.0   
498         4.0          4.0          4.0          4.0          4.0   
499         4.0          4.0          4.0          4.0          4.0   

     Unnamed: 34  Unnamed: 35  
0            4.0          4.0  
1            4.0          4.0  
2            4.0          4.0  
3            4.0          4.0  
4            4.0          4.0  
..           ...          ...  
495          4.0          4.0  
496          4.0          4.0  
497          4.0          4.0  
498          4.0          4.0  
499          4.0          4.0  

[500 rows x 36 columns]

In [150]:
df2.fillna(value=df['profit'].mean())

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0    1006032852       2114       2274      2491      2660      2727      2887   
1    1212192089        151        275       570       750       813      1066   
2    1190380701          6          7         7         7         8         8   
3     966005983       1005       1128      1326      1455      1507      1621   
4     876339541        996       1175      1304      1396      1432      1559   
..          ...        ...        ...       ...       ...       ...       ...   
495   713391965          0          0         0       560       554       544   
496   722565148          0          0         0       875       866       861   
497   532874347          0          0         0       734       728       726   
498   655464934          0          0         0       254       259       261   
499   919930954          0          0         0       538       545       558   

     10/9/2013 12/9/2013 14/9/2013  ... Unnamed: 26 Unnamed: 27 Unnamed: 28  \
0         2930      3119      3204  ...    28.64174    28.64174    28.64174   
1         1164      1558      1756  ...    28.64174    28.64174    28.64174   
2            9        10        10  ...    28.64174    28.64174    28.64174   
3         1637      1723      1746  ...    28.64174    28.64174    28.64174   
4         1570      1638      1655  ...    28.64174    28.64174    28.64174   
..         ...       ...       ...  ...         ...         ...         ...   
495        537       525       519  ...    28.64174    28.64174    28.64174   
496        854       850       844  ...    28.64174    28.64174    28.64174   
497        715       694       690  ...    28.64174    28.64174    28.64174   
498        263       268       270  ...    28.64174    28.64174    28.64174   
499        563       578       585  ...    28.64174    28.64174    28.64174   

    Unnamed: 29  Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  \
0      28.64174     28.64174     28.64174     28.64174     28.64174   
1      28.64174     28.64174     28.64174     28.64174     28.64174   
2      28.64174     28.64174     28.64174     28.64174     28.64174   
3      28.64174     28.64174     28.64174     28.64174     28.64174   
4      28.64174     28.64174     28.64174     28.64174     28.64174   
..          ...          ...          ...          ...          ...   
495    28.64174     28.64174     28.64174     28.64174     28.64174   
496    28.64174     28.64174     28.64174     28.64174     28.64174   
497    28.64174     28.64174     28.64174     28.64174     28.64174   
498    28.64174     28.64174     28.64174     28.64174     28.64174   
499    28.64174     28.64174     28.64174     28.64174     28.64174   

     Unnamed: 34  Unnamed: 35  
0       28.64174     28.64174  
1       28.64174     28.64174  
2       28.64174     28.64174  
3       28.64174     28.64174  
4       28.64174     28.64174  
..           ...          ...  
495     28.64174     28.64174  
496     28.64174     28.64174  
497     28.64174     28.64174  
498     28.64174     28.64174  
499     28.64174     28.64174  

[500 rows x 36 columns]

In [151]:
df2['Unnamed: 35'].fillna(value=df['profit'].mean())

0      28.64174
1      28.64174
2      28.64174
3      28.64174
4      28.64174
         ...   
495    28.64174
496    28.64174
497    28.64174
498    28.64174
499    28.64174
Name: Unnamed: 35, Length: 500, dtype: float64

In [152]:
df2.head()

Dress_ID  29/8/2013  31/8/2013  2/9/2013  4/9/2013  6/9/2013  8/9/2013  \
0  1006032852       2114       2274      2491      2660      2727      2887   
1  1212192089        151        275       570       750       813      1066   
2  1190380701          6          7         7         7         8         8   
3   966005983       1005       1128      1326      1455      1507      1621   
4   876339541        996       1175      1304      1396      1432      1559   

   10/9/2013 12/9/2013 14/9/2013  ... Unnamed: 26 Unnamed: 27 Unnamed: 28  \
0       2930      3119      3204  ...         NaN         NaN         NaN   
1       1164      1558      1756  ...         NaN         NaN         NaN   
2          9        10        10  ...         NaN         NaN         NaN   
3       1637      1723      1746  ...         NaN         NaN         NaN   
4       1570      1638      1655  ...         NaN         NaN         NaN   

  Unnamed: 29  Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 34  Unnamed: 35  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN  

[5 rows x 36 columns]

In [156]:
df.groupby('product_id')['profit'].mean()

product_id
FUR-ADV-10000002     30.19500
FUR-ADV-10000108      1.12000
FUR-ADV-10000183    -81.46725
FUR-ADV-10000188      0.84000
FUR-ADV-10000190    104.46000
                      ...    
TEC-STA-10004181     30.48200
TEC-STA-10004536     10.83600
TEC-STA-10004542     22.79040
TEC-STA-10004834    -72.78150
TEC-STA-10004927     15.48000
Name: profit, Length: 10292, dtype: float64

In [157]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ...  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...   
3  Home Office        Stockholm     Sweden      EU    North  ...   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...   

                  product_name sales quantity discount   profit  \
0          Tenex Lockers, Blue   408        2      0.0  106.140   
1     Acme Trimmer, High Speed   120        3      0.1   36.036   
2      Tenex Box, Single Width    66        4      0.0   29.640   
3  Enermax Note Cards, Premium    45        3      0.5  -26.055   
4   Eldon Light Bulb, Duo Pack   114        5      0.1   37.770   

   shipping_cost  order_priority  new_order_date  order_month cost_to_company  
0          35.46          Medium      2011-01-01            1           35.46  
1           9.72          Medium      2011-01-01            1            9.82  
2           8.17            High      2011-01-01            1            8.17  
3           4.82            High      2011-01-01            1            5.32  
4           4.70          Medium      2011-01-01            1            4.80  

[5 rows x 23 columns]

In [159]:
df.groupby('order_priority')['profit'].mean()

order_priority
Critical    31.994327
High        27.119122
Low         24.197958
Medium      29.361729
Name: profit, dtype: float64

In [161]:
df.dtypes

order_id                   object
order_date                 object
ship_date                  object
ship_mode                  object
customer_name              object
segment                    object
state                      object
country                    object
market                     object
region                     object
product_id                 object
category                   object
sub_category               object
product_name               object
sales                      object
quantity                    int64
discount                  float64
profit                    float64
shipping_cost             float64
order_priority             object
new_order_date     datetime64[ns]
order_month                 int64
cost_to_company           float64
dtype: object

In [162]:
df['order_year'] = df['new_order_date'].dt.year

In [163]:
df.head()

order_id order_date ship_date       ship_mode    customer_name  \
0     AG-2011-2040   1/1/2011  1/6/2011  Standard Class  Toby Braunhardt   
1    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   
2     HU-2011-1220   1/1/2011  1/5/2011    Second Class    Annie Thurman   
3  IT-2011-3647632   1/1/2011  1/5/2011    Second Class     Eugene Moren   
4    IN-2011-47883   1/1/2011  1/8/2011  Standard Class      Joseph Holt   

       segment            state    country  market   region  ... sales  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...   408   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...   120   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...    66   
3  Home Office        Stockholm     Sweden      EU    North  ...    45   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...   114   

  quantity discount   profit shipping_cost  order_priority  new_order_date  \
0        2      0.0  106.140         35.46          Medium      2011-01-01   
1        3      0.1   36.036          9.72          Medium      2011-01-01   
2        4      0.0   29.640          8.17            High      2011-01-01   
3        3      0.5  -26.055          4.82            High      2011-01-01   
4        5      0.1   37.770          4.70          Medium      2011-01-01   

   order_month  cost_to_company order_year  
0            1            35.46       2011  
1            1             9.82       2011  
2            1             8.17       2011  
3            1             5.32       2011  
4            1             4.80       2011  

[5 rows x 24 columns]

In [164]:
df.groupby('order_year')['profit'].mean()

order_year
2011    27.666238
2012    28.043722
2013    29.604519
2014    28.758540
Name: profit, dtype: float64

In [175]:
def replace_coma(x):
    x = x.replace(',','')
    x = int(x)
    return x

In [177]:
#df['sales'].apply(replace_coma)

In [173]:
df['sales'].unique()

array(['408', '120', '66', ..., '1,763', '1,821', '1,831'], dtype=object)

In [178]:
df['sales'] = df.sales.str.replace(',','').astype(int)

In [179]:
df.dtypes

order_id                   object
order_date                 object
ship_date                  object
ship_mode                  object
customer_name              object
segment                    object
state                      object
country                    object
market                     object
region                     object
product_id                 object
category                   object
sub_category               object
product_name               object
sales                       int32
quantity                    int64
discount                  float64
profit                    float64
shipping_cost             float64
order_priority             object
new_order_date     datetime64[ns]
order_month                 int64
cost_to_company           float64
order_year                  int64
dtype: object

In [180]:
df.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority',
       'new_order_date', 'order_month', 'cost_to_company', 'order_year'],
      dtype='object')

In [183]:
df.groupby('order_year')['quantity'].sum()

order_year
2011    31443
2012    38111
2013    48136
2014    60622
Name: quantity, dtype: int64

In [184]:
df.groupby('order_year')['sales'].mean()

order_year
2011    251.112581
2012    244.252235
2013    246.819335
2014    245.282129
Name: sales, dtype: float64

In [185]:
df.groupby('order_year')['profit'].mean()

order_year
2011    27.666238
2012    28.043722
2013    29.604519
2014    28.758540
Name: profit, dtype: float64

# Challenges

1. find out their av. rating on weekly basis, keeping in mind two days leave
2. total working days for each agent
3. total query taken, 
4. total feedback receved
5. agent name who have av rating between 3.5 to 4
6. agent name who have rating less than 3.5
7. agent name who have rating > 4.5
8. how many feedback agent have recieved > 4.5 average
9. average weekly response time for each agent 
10. average weekely resolution time for each agents 
11. list of all agents name 
12. percentage of chat on which they have received a feedback 
13. Total contributation hour for each and every agents weekly basis 
14. total percentage of active hour for a month 

sunny.savita@ineuron.ai

deadline today : 12am today

1 .Find out there avarage rating on weekly basis keep this in a mind that they take two days of leave
2 .Total working days for each agents 
3. Total query that you hvae taken 
4. total Feedback that you have received 
5. a agent name who have average rating between 3.5 to 4 
6 . Agent name who have rating lesss then 3.5 
7 . agent name who have rating more then 4.5 
8 . how many feedaback agents have received more then 4.5 average
9 . average weekly response time for each agent 
10 . average weekely resolution time for each agents 
11 . list of all agents name 
12 . percentage of chat on which they have received a feedback 
13 . Total contributation hour for each and every agents weekly basis 
14. total percentage of active hour for a month

sunny.savita@ineuron.ai ,sudhanshu@ineuron.ai